Enhance supported patterns to include ** matching multilevel of subdomains

pjablonski123 opened the pull request:
Adding ** to match substrings as a prefix including subdomains.


<!-- Description of change -->
```
So far subdomains required a definition of all subdomains by adding a chain of star and dot.

Example:
rules:
  dns:
    - matchPattern: '*.cluster.local'
    - matchPattern: '*.*.cluster.local'
    - matchPattern: '*.*.*.cluster.local'
    - matchPattern: '*.*.*.*.cluster.local'
    - matchPattern: '*.*.*.*.*.cluster.local'
    - matchPattern: '*.*.*.*.*.*.cluster.local'
    - matchPattern: '*.*.*.*.*.*.*.cluster.local'
    - matchPattern: '*.*.*.*.*.*.*.*.cluster.local'
    - matchPattern: '*.*.*.internal'
    - matchPattern: '*.*.*.*.internal'
    - matchPattern: '*.*.*.*.*.internal'
    - matchPattern: '*.*.*.*.*.*.internal'
    - matchPattern: '*.*.*.*.*.*.*.internal'
    - matchPattern: '*.*.*.*.*.*.*.*.internal'
    - matchPattern: '*.*.*.*.*.*.*.*.*.internal'
    - matchPattern: '*.*.internal.cloudapp.net'
    - matchPattern: '*.*.*.internal.cloudapp.net'
    - matchPattern: '*.*.*.*.internal.cloudapp.net'
    - matchPattern: '*.*.*.*.*.internal.cloudapp.net'
    - matchPattern: '*.*.*.*.*.*.internal.cloudapp.net'
    - matchPattern: '*.*.*.*.*.*.*.internal.cloudapp.net'
    - matchPattern: '*.*.*.*.*.*.*.*.internal.cloudapp.net'

A feature enhancement introduces wildcards where ** would represent 1 or more DNS subdomains. That would allow them to compress the above policy to:

rules:
  dns:
    - matchPattern: '**.cluster.local'
    - matchPattern: '**.internal'
    - matchPattern: '**.internal.cloudapp.net'

Th code modification does not change the old behaviour if someone wants to list subdomains explicitly.

For example the policy:
apiVersion: cilium.io/v2
kind: CiliumNetworkPolicy
metadata:
  name: fqdn-policy
spec:
  endpointSelector: {}
  egress:
    - toFQDNs:
      - matchPattern: "*pl"
      - matchPattern: "*.io"
      - matchPattern: "**com"
      - matchPattern: "**.amazonaws.com"
    - toPorts:
      - ports:
         - port: "53"
           protocol: UDP
        rules:
          dns:
            - matchPattern: "*"
         
produces the following result:
Forbids domain.pl as expected by the old behaviour.
Allows cilium.io as expected by the old behaviour.
Allows google.com as expected by a new behaviour.
Allows elasticbeanstalk-eu-west-1-123456789000.s3-eu-west-1.amazonaws.com as expected by a new behaviour.
```
Release note:
```release-note
Enhance supported patterns to include ** matching multilevel of subdomains.

Instead of the rules with listed subdomains:

rules:
  dns:
    - matchPattern: '*.cluster.local'
    - matchPattern: '*.*.cluster.local'
    - matchPattern: '*.*.*.cluster.local'
    - matchPattern: '*.*.*.*.cluster.local'
    - matchPattern: '*.*.*.*.*.cluster.local'
    - matchPattern: '*.*.*.*.*.*.cluster.local'
    - matchPattern: '*.*.*.*.*.*.*.cluster.local'
    - matchPattern: '*.*.*.*.*.*.*.*.cluster.local'

the following policy can be used:

rules:
  dns:
    - matchPattern: '**.cluster.local'
 
```
Signed-off-by: Piotr Jablonski <piotr.jablonski@isovalent.com>

Fixes: https://github.com/cilium/cilium/issues/22081

joestringer reviewed (dismissed):
I didn't review the code but I just wanted to reshare the concern from the issue which we need to convince ourselves is OK: https://github.com/cilium/cilium/issues/22081#issuecomment-1314559494

bimmlerd commented on pkg/fqdn/matchpattern/matchpattern.go:
```diff
@@ -70,6 +70,11 @@ func ToAnchoredRegexp(pattern string) string {
 	pattern = strings.ToLower(pattern)
 
+	// handle the ** match-all case.
+	if pattern == "**" {
```
If we end up returning the same as for the `pattern == "*"` case, I'd make it 

`if pattern == "*" || pattern == "**" {` to avoid the duplication.

pjablonski123 replied:
Good point, thank you!

bimmlerd commented on pkg/fqdn/matchpattern/matchpattern.go:
```diff
@@ -70,6 +70,11 @@ func ToAnchoredRegexp(pattern string) string {
 	pattern = strings.TrimSpace(pattern)
```
function doc comment should learn about `"**"` too

pjablonski123 replied:
Added.

bimmlerd commented on pkg/fqdn/matchpattern/matchpattern.go:
```diff
@@ -88,6 +93,12 @@ func ToAnchoredRegexp(pattern string) string {
 func ToUnAnchoredRegexp(pattern string) string {
```
function doc comment should learn about `"**"` too

pjablonski123 replied:
Added

bimmlerd commented on pkg/fqdn/matchpattern/matchpattern.go:
```diff
@@ -88,6 +93,12 @@ func ToAnchoredRegexp(pattern string) string {
 	pattern = strings.ToLower(pattern)
+
+	// handle the ** match-all case.
+	if pattern == "**" {
```
as above, `if pattern == "*" || pattern == "*"`

pjablonski123 replied:
Corrected.

bimmlerd commented on pkg/fqdn/matchpattern/matchpattern_test.go:
```diff
@@ -58,6 +64,8 @@ func (ts *MatchPatternTestSuite) TestUnAnchoredMatchPatternREConversion(c *C) {
 // *.cilium.io. matches anysub.cilium.io. but not cilium.io.
 // *cilium.io. matches  anysub.cilium.io. and cilium.io.
 // *.ci*.io. matches anysub.cilium.io. anysub.ci.io., anysub.ciliumandmore.io. but not cilium.io.
+// **.cilium.io. matches  anysub.cilium.io. andmomre subdomains.
```
typo, probably 
```suggestion
// **.cilium.io. matches anysub.cilium.io. and more subdomains.
```

pjablonski123 replied:
Fixed.

bimmlerd commented on pkg/fqdn/matchpattern/matchpattern_test.go:
```diff
@@ -58,6 +64,8 @@ func (ts *MatchPatternTestSuite) TestUnAnchoredMatchPatternREConversion(c *C) {
 // *cilium.io. matches  anysub.cilium.io. and cilium.io.
 // *.ci*.io. matches anysub.cilium.io. anysub.ci.io., anysub.ciliumandmore.io. but not cilium.io.
+// **.cilium.io. matches  anysub.cilium.io. andmomre subdomains.
+// **.ci*.io. matches anysub.cilium.io. anysub.ci.io. and more subdomains anysub.ciliumandmore.io. but not cilium.io.
```
It seems relevant to call out that it also matches `some.more.ci.io`, or did I misunderstand the point of the PR?

pjablonski123 replied:
Added your suggestion, thanks! There are many options to included them all here.

bimmlerd commented on pkg/fqdn/matchpattern/matchpattern_test.go:
```diff
@@ -117,6 +182,7 @@ func (ts *MatchPatternTestSuite) TestAnchoredMatchPatternMatching(c *C) {
 // TestMatchPatternSanitize tests that Sanitize handles any special cases
 func (ts *MatchPatternTestSuite) TestMatchPatternSanitize(c *C) {
 	for source, target := range map[string]string{
+		"**":    "**.",
```
also `"**."` to check that it's not overeager?

pjablonski123 replied:
Added.

bimmlerd commented on pkg/fqdn/matchpattern/matchpattern_test.go:
```diff
@@ -27,10 +27,13 @@ var _ = Suite(&MatchPatternTestSuite{})
+		"_sub.**.io.":   "^_sub[.](" + allowedDNSCharsREGroup + "+[.])+io[.]$",
+		"**":            "(^(" + allowedDNSCharsREGroup + "+[.])+$)|(^[.]$)",
+		"*":             "(^(" + allowedDNSCharsREGroup + "+[.])+$)|(^[.]$)",
+		".":             "^[.]$",
```
some more ideas for test cases (admittedly only some of these related to the PR)

```
"**."
"*."
"**.cilium.**.io"
"**cilium.**.io"
"cilium.**.**.io"
"**.**.cilium.io"
"**.**."
"**.**"
"cilium.**.**"
"cilium.**.**"
```

pjablonski123 replied:
A single * did not have such extensive tests, but I added all of them.

pjablonski123 replied:
Thank you for suggestions!

bimmlerd reviewed (changes requested):
Question: do we handle `**.**.cilium.io` and `**.cilium.**.io`, and if so, how? 

Second question: should `cilium.**` be accepted?

Third question: What's correct for `**cilium.io`? Should it behave like it seems to, currently? I'd expect `a.b.cilium.io.` to match if `a.cilium.io.` does.

```
raw: "**cilium.io"
regex: ^[-a-zA-Z0-9_]+[.][-a-zA-Z0-9_]*cilium[.]io[.]$
yes:	"a.cilium.io."
no:	"a.b.cilium.io."
no:	"a.b.c.cilium.io."
```

Oh and are there docs on how this should behave? The semantics aren't all that clear to me, tbh.

gentoo-root commented on pkg/fqdn/matchpattern/matchpattern.go:
```diff
@@ -100,8 +111,15 @@ func escapeRegexpCharacters(pattern string) string {
 	// base case. * becomes .*, but only for DNS valid characters
 	// NOTE: this only works because the case above does not leave the *
 	pattern = strings.Replace(pattern, "*", allowedDNSCharsREGroup+"*", -1)
+
```
This `Replace` will match on `*` added in the previous `Replace`. It doesn't look intended, is it?

pjablonski123 replied:
It seems to be ok, as the ** is resolved to the chain of allowedDNSCharsREGroup and [.]

gentoo-root replied:
That's not what I read a few lines above:

```
	// ** becomes [-a-zA-Z0-9_.]+[.]*
	pattern = strings.Replace(pattern, "**", allowedDNSCharsREGroup+"+[.]*", -1)
```

What about that `*` in the end of the replacement string?

pjablonski123 replied:
Right, the comment is a leftover and it is related to a previous state before "**" was added. So the asterisk in the pattern:

`pattern = strings.Replace(pattern, "**", allowedDNSCharsREGroup+"+[.]*", -1)`

is resolved by the following line:

`pattern = strings.Replace(pattern, "*", allowedDNSCharsREGroup+"*", -1)`

which is how subdomains can be stacking on top of the main domain.

pjablonski123 replied:
I will change the comments, thank you for catching this one.

gentoo-root replied:
> is resolved by the following line

Oh, so this is intentional? Adding some comments will definitely help understanding this logic :)

So, basically, the conversions are:

```
**. -> ([-a-zA-Z0-9_.]+[.])+
 ** -> [-a-zA-Z0-9_.]+[.][-a-zA-Z0-9_.]*
  * -> [-a-zA-Z0-9_.]*
```

I'm still curious about two things:
1. Why is a dot included into `[-a-zA-Z0-9_.]`?
2. Why does `**` force one single dot?

gentoo-root replied:
Oh, I guess the answer for (1) is that there isn't actually a dot there, it's only in the comment. The comment should be fixed, otherwise it's fine.

For (2), I see there is a discussion below on the semantics of `**` without a following dot.

pjablonski123 replied:
(2) Yes, the presence of dot was the most difficult one I did not want to cover all possible corner cases. Hence some limitations like with `**cilium.io` which matches `a.cilium.io`, but not `a.b.cilium.io`. 
The recommended form with dot which is this one: `**.cilium.io` matches all subdomains.

pjablonski123 replied:
(1) Correct! Thanks for your remarks, too many changes along the way, I will fix the comment.

tklauser reviewed (changes requested):
@pjablonski123 Please also add some documentation to `Documentation/security/policy/language.rst` describing the new `**` matching behavior.

bimmlerd reviewed (dismissed):
Awesome progress, this is much better already.

However, if we don't know what to specify for the behaviour of `**cilium.io`, then we should reject it for now. Otherwise, we are bound by the (implementation-defined) behaviour and will not be able to change it in the future, as that would be a breaking change (essentially https://www.hyrumslaw.com/).

tklauser commented on Documentation/security/policy/language.rst:
```diff
@@ -587,6 +594,36 @@ Example
+the ``**`` and the chain of ``*.`` wildcards in particular could have far-reaching
+effects. Before implementing policies which use such wildcards, users should 
+evaluate their own threat model and the likelihood of a target domain or 
+sub-domain being configured or hijacked to exfiltrate information from the Cilium 
```
The documentation CI workflow is currently failing because the term `exfiltrate` is not part of the spelling list: https://github.com/cilium/cilium/actions/runs/7047078694/job/19215007236?pr=28690

You can fix this by either using a different term or by adding it as an exception to `Documentation/spelling_wordlist.txt`.

pjablonski123 replied:
Exfiltrate is a common security word describing of what may happen with data. I will add an exception.

pjablonski123 committed:
Enhance supported patterns to include ** matching multilevel of subdomains.

pkg/fqdn: Adding ** matchpattern

The ** in matchpattern allows a short notation of multilevel subdomains

Instead of the rules with listed subdomains:

rules:
  dns:
    - matchPattern: '*.cluster.local'
    - matchPattern: '*.*.cluster.local'
    - matchPattern: '*.*.*.cluster.local'
    - matchPattern: '*.*.*.*.cluster.local'
    - matchPattern: '*.*.*.*.*.cluster.local'
    - matchPattern: '*.*.*.*.*.*.cluster.local'
    - matchPattern: '*.*.*.*.*.*.*.cluster.local'
    - matchPattern: '*.*.*.*.*.*.*.*.cluster.local'

egress:
  - toFQDNs:
    - matchPattern: '*.cilium.io'
    - matchPattern: '*.*.cilium.io'
    - matchPattern: '*.*.*.cilium.io'
    - matchPattern: '*.*.*.*.cilium.io'
    - matchPattern: '*.*.*.*.*.cilium.io'
    - matchPattern: '*.*.*.*.*.*.cilium.io'
    - matchPattern: '*.*.*.*.*.*.*.cilium.io'
    - matchPattern: '*.*.*.*.*.*.*.*.cilium.io'

the following policy can be used in a simpler form:

rules:
  dns:
    - matchPattern: '**.cluster.local'

egress:
  - toFQDNs:
    - matchPattern: '**.cilium.io'

Changelog:
Modified the code and added tests.
Updated language.rst.
Updated matchpattern.go.
Added and updated doc files.
Updated language.rst with a security note.
Double asterisk is allowed at the beginning of the domain only.

Signed-off-by: Piotr Jablonski <pijablon@gmail.com>

```release-note
The ** in matchpattern allows a short notation of multilevel subdomains,
eg. **.cilium.io matches a.cilium.io, a.b.cilium.io, a.b.c.cilium.io, etc.

This feature can break current configuration if a double asterisk was used inside of FQDN, eg. cil**.io
Before this enhancement two or more directly adjacent asterisks meant the same - a single asterisk, eg. cil**.io == cil*.io
The correct usage of the asterisk before this enhancement should be a single presence between other characters, eg. c*m.io, *.cil*.io
With this enhancement the correct usage of the single asterisk does not change
A double asterisk must be at the beginning of the FQDN string only, eg. **.cilium.io
The single and double asterisk can be present in the same FQDN if they are separated by a dot directly or indirectly, eg. **.*lium.io, **.cili*.io
```

gentoo-root reviewed (approved):
Looks good to me, assuming that folks agreed on the semantics of `"**xyz"`.
